[GStreamer][WebRTC] Move network access to NetworkProcess

philn opened the pull request:
#### 53277244d7bb90735d32204243f5bcacf9cdb742
<pre>
[GStreamer][WebRTC] Move network access to NetworkProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=244002">https://bugs.webkit.org/show_bug.cgi?id=244002</a>

Reviewed by Xabier Rodriguez-Calvar and Adrian Perez de Castro.

This patch introduces a build-time dependency on librice for the GTK port. The USE_LIBRICE CMake
option can be set to FALSE if this dependency is not desired. It is FALSE by default on WPE.

When USE_LIBRICE is enabled, the WebRTC network handling is now deferred to the NetworkProcess,
bi-directional communication is performed with the WebProcess which makes use of a custom
GstWebRTCICE implementation. Runtime to the previous un-sandboxed libnice-based implementation can
be done by setting the WEBKIT_GST_DISABLE_WEBRTC_NETWORK_SANDBOX=1 environment variable.

Canonical link: <a href="https://commits.webkit.org/303623@main">https://commits.webkit.org/303623@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d68367c15f0a99f7821d6eec7b98ca389f58c467

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133049 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5550 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44147 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140585 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85084 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134919 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6027 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5413 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/101748 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69095 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135995 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119227 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82547 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/132398 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1717 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83820 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125119 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37342 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143238 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/131557 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5219 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37922 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/110127 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5301 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/4477 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110309 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27959 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4017 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115485 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58820 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5273 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33836 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/164525 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5117 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68725 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/42882 "Build was cancelled. Recent messages:") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5363 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5231 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.h:
```diff
@@ -0,0 +1,120 @@
+public:
+    static Ref<RiceBackendClient> create() { return adoptRef(*new RiceBackendClient); }
+    void ref() const { RefCounted::ref(); }
+    void deref() const { RefCounted::deref(); }
```
Why do we need this?

philn replied:
That's standard procedure for RefCounted sub-classes.

philn replied:
But yeah it's in the base class, I'll check this again.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.h:
```diff
@@ -0,0 +1,120 @@
+
+    using IncomingDataCallback = WTF::Function<void(unsigned, RTCIceProtocol, String, String, std::span<const uint8_t>&&)>;
+    void setIncomingDataCallback(IncomingDataCallback&& callback) { m_incomingDataCallback = WTFMove(callback); }
+    void notifyIncomingData(unsigned streamId, RTCIceProtocol protocol, String from, String to, std::span<const uint8_t>&& data) { m_incomingDataCallback(streamId, protocol, from, to, WTFMove(data)); }
```
The String should either be const& or &&, right? Am I missing anything?

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.h:
```diff
@@ -0,0 +1,120 @@
+class RiceBackend : public Identified<RiceBackendIdentifier> {
+    WTF_MAKE_NONCOPYABLE(RiceBackend);
+public:
+    void ref() { refRiceBackend(); }
+    void deref() { derefRiceBackend(); }
```
ditto

philn replied:
That's needed.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.h:
```diff
@@ -0,0 +1,120 @@
+    using ResolveAddressCallback = Function<void(ExceptionOr<String>&&)>;
+    virtual void resolveAddress(const String&, ResolveAddressCallback&&) = 0;
+
+    virtual void send(unsigned, RTCIceProtocol, String, String, std::span<const uint8_t>) = 0;
```
ditto

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.h:
```diff
@@ -0,0 +1,120 @@
+
+GstWebRTCICETransport *webkitGstWebRTCIceAgentCreateTransport(WebKitGstIceAgent*, GThreadSafeWeakPtr<WebKitGstIceStream>&&, WebCore::RTCIceComponent);
+
+void webkitGstWebRTCIceAgentSend(WebKitGstIceAgent*, unsigned, WebCore::RTCIceProtocol, String from, String to, std::span<const uint8_t>);
```
ditto

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,768 @@
+{
+    auto backend = WEBKIT_GST_WEBRTC_ICE_BACKEND(ice);
+    backend->priv->stunServer = String::fromUTF8(uri);
+    GST_INFO_OBJECT(ice, "Setting STUN server address to %s", backend->priv->stunServer.utf8().data());
```
Only if someone were writing a CStringView class so that you could wrap the URI, then transform into a String to keep it and eventually, using the original CStringView to recover the contents with no extra copy...

Anyway, here you can avoid using the copy of utf8() by using the original uri parameter.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,768 @@
+    const auto& host = url.host();
+    auto port = url.port().value_or(3478);
+
+    backend->priv->iceBackend->resolveAddress(host.toString(), [weakAgent = GThreadSafeWeakPtr(backend), port](auto&& result) mutable {
```
Can we please properly type result? IMHO, auto should be used then the type is also quickly deductible from reading the code, which is not in this case.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,768 @@
+    auto url = *validationResult;
+    auto wasAdded = backend->priv->turnServers.add(url).isNewEntry;
+    if (!wasAdded) {
+        GST_WARNING_OBJECT(ice, "%s was already registered, no need to add it again", uri);
```
If this can normally happen, it should not be a warning, just debug would be ok.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,768 @@
+        GST_ERROR_OBJECT(ice, "Error validating TURN URI: %s", validationResult.error().data.utf8().data());
+        return;
+    }
+    backend->priv->turnServer = String::fromUTF8(uri);
```
You're creating the string twice.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,768 @@
+            return;
+        }
+
+        auto newCandidateSdp = makeString(WTFMove(prefix), ' ', result.returnValue(), ' ', WTFMove(suffix));
```
Why are you moving something when you make a string. Isn't this just taking a const String& and forcing that move is going to be no-op?

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,768 @@
+
+static gboolean webkitGstWebRTCIceAgentSetLocalCredentials(GstWebRTCICE*, GstWebRTCICEStream* stream, const gchar* ufrag, const gchar* pwd)
+{
+    webkitGstWebRTCIceStreamSetLocalCredentials(WEBKIT_GST_WEBRTC_ICE_STREAM(stream), String::fromLatin1(ufrag), String::fromLatin1(pwd));
```
I would assume passwords to be UTF8 and maybe the ufrag as well

philn replied:
No, see the definition of `ice-char` in https://datatracker.ietf.org/doc/html/rfc8839.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,768 @@
+
+static gboolean webkitGstWebRTCIceAgentSetRemoteCredentials(GstWebRTCICE*, GstWebRTCICEStream* stream, const gchar* ufrag, const gchar* pwd)
+{
+    webkitGstWebRTCIceStreamSetRemoteCredentials(WEBKIT_GST_WEBRTC_ICE_STREAM(stream), String::fromLatin1(ufrag), String::fromLatin1(pwd));
```
ditto

philn replied:
ditto

aperezdc commented on Source/cmake/FindLibRiceProto.cmake:
```diff
@@ -0,0 +1,67 @@
+    VERSION_VAR LibRiceProto_VERSION
+)
+
+if (LibRiceProto_LIBRARY AND NOT TARGET LibRice::Proto)
```
I think we should have a single `FindRice.cmake` module, which accepts indicating which components to search for. This is what we do for other libraries like GLib or WebP. Then we would use:

```cmake
find_package(Rice 0.1.1 COMPONENTS Io Proto)
```

philn replied:
Done, but as I'm not really CMake expert, can you please take another look?

aperezdc commented on Source/cmake/GStreamerChecks.cmake:
```diff
@@ -65,6 +65,14 @@ if (ENABLE_MEDIA_STREAM AND ENABLE_WEB_RTC)
+            find_package(LibRiceProto 0.1.1 REQUIRED)
+            if (NOT LibRiceIo_FOUND OR NOT LibRiceProto_FOUND)
+                message(FATAL_ERROR "librice-{io,proto} is needed for USE_LIBRICE.")
+            endif ()
```
We do not need to manually produce an error when passing `REQUIRED` to the `find_package()` command; CMake will notify itself about the error. Sometimes we skip the `REQUIRED` flag in order to give a better error message and hinting about an `USE_*` or `ENABLE_*` option that can be toggled, which I think is the case here.


```suggestion
            find_package(LibRiceIo 0.1.1)
            find_package(LibRiceProto 0.1.1)
            if (NOT LibRiceIo_FOUND OR NOT LibRiceProto_FOUND)
                message(FATAL_ERROR "librice-{io,proto} is needed for USE_LIBRICE.")
            endif ()
```

aperezdc reviewed (commented):
I only took a look at the CMake bits.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,780 @@
+    std::array<RiceTransportType, 4> relays = { static_cast<RiceTransportType>(0), };
+    unsigned nRelay = 0;
+    bool isTurns = url.protocolIs("turns"_s);
+    StringView transport;
```
CStringView

philn replied:
CStringView has no =operator.

calvaris replied:
In the code I am seeing below it will use the constructor implicitly. It should be no problem. What am I missing?

calvaris replied:
You didn't change this.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,780 @@
+
+    RELEASE_ASSERT(url.port());
+    auto port = url.port().value();
+    agent->priv->iceBackend->resolveAddress(url.host().toString(), [weakAgent = GThreadSafeWeakPtr(agent), isTurns, port, nRelay, user = url.user(), password = url.password(), relays = WTFMove(relays)](auto&& result) mutable {
```
de-auto this result

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,780 @@
+static gchar* webkitGstWebRTCIceAgentGetTurnServer(GstWebRTCICE* ice)
+{
+    auto backend = WEBKIT_GST_WEBRTC_ICE_BACKEND(ice);
+    return g_strdup(backend->priv->turnServer.utf8().data());
```
Note to self: do something to change the idiom of this dups when I finish writing GMallocString.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,780 @@
+    String suffix;
+};
+
+static Expected<CandidateAddress, ExceptionData> getCandidateAddress(StringView candidate)
```
We can probably use CStringView here.

philn replied:
It would complexify the code for no added benefit.

calvaris replied:
In what way? The benefit is using a lighter class, to begin with.

calvaris replied:
Please.

philn replied:
> In what way? 

Because the StringView API is actually used in this function. How do you get a sub string from a CStringView and split that? Also `WTF::startsWith()` isn't very idiomatic.

calvaris replied:
Ok, leave it as it is then.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,780 @@
+        return "tls"_s;
+
+    ASSERT(url.protocolIs("turn"_s));
+    StringView transport;
```
CStringView

philn replied:
CStringView has no =operator.

calvaris replied:
ditto

calvaris replied:
Please.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,780 @@
+#endif
+}
+
+WebKitGstIceAgent* webkitGstWebRTCCreateIceAgent(StringView name, ScriptExecutionContext* context)
```
CStringView

philn replied:
Actually a `const String&` makes more sense here.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,780 @@
+        GUniquePtr<char> sdpCandidate(rice_candidate_to_sdp_string(&candidate.gathered.candidate));
+        auto sdp = String::fromUTF8(sdpCandidate.get());
+        ASSERT(sdp.startsWith("a="_s));
+        auto strippedSdp = sdp.substring(2);
```
You can use CStringView here even if you have to convert it to a CString after subspanning to make it null terminated.

calvaris replied:
Though now I am thinking that if you use the spanIncludingNullTerminator you can save a copy! It would be nice to have CStringView constructor from spans. Please add a fixme so that I can check it later.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceStream.cpp:
```diff
@@ -0,0 +1,312 @@
+
+void webkitGstWebRTCIceStreamSetLocalCredentials(WebKitGstIceStream* stream, const String& ufrag, const String& pwd)
+{
+    GUniquePtr<RiceCredentials> credentials(rice_credentials_new(ufrag.ascii().data(), pwd.ascii().data()));
```
use utf8 for ufrag and password.

philn replied:
Not needed.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceStream.cpp:
```diff
@@ -0,0 +1,312 @@
+
+void webkitGstWebRTCIceStreamSetRemoteCredentials(WebKitGstIceStream* stream, const String& ufrag, const String& pwd)
+{
+    GUniquePtr<RiceCredentials> credentials(rice_credentials_new(ufrag.ascii().data(), pwd.ascii().data()));
```
ditto

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -169,10 +176,26 @@ bool GStreamerMediaEndpoint::initializePipeline()
+    nPipeline++;
+
+#if USE(LIBRICE) && GST_CHECK_VERSION(1, 20, 0)
+    auto disableNetworkSandbox = StringView::fromLatin1(g_getenv("WEBKIT_GST_DISABLE_WEBRTC_NETWORK_SANDBOX"));
```
CStringView

calvaris commented on Source/WebKit/NetworkProcess/webrtc/rice/RiceBackend.cpp:
```diff
@@ -0,0 +1,359 @@
+void RiceBackend::notifyIncomingData(unsigned streamId, WebCore::RTCIceProtocol protocol, String&& from, String&& to, std::span<const uint8_t>&& data)
+{
+    callOnMainRunLoopAndWait([&, streamId, protocol, data = WTFMove(data), from = WTFMove(from), to = WTFMove(to)] {
+        messageSenderConnection()->send(Messages::RiceBackendProxy::NotifyIncomingData { streamId, protocol, from, to, data }, messageSenderDestinationID());
```
Don't you need a mutable and moves here for the strings

philn replied:
mutable yes, but the strings can't be moved (generated constructor for this uses const String&).

calvaris replied:
If the generated constructor is using const& then you won't need the mutable either because they won't be moved.

calvaris commented on Source/WebKit/NetworkProcess/webrtc/rice/RiceBackend.h:
```diff
@@ -0,0 +1,114 @@
+    static void initialize(NetworkConnectionToWebProcess&, WebKit::WebPageProxyIdentifier&&, CompletionHandler<void(RefPtr<RiceBackend>&&)>&&);
+    ~RiceBackend();
+
+    void ref() const final { RefCounted::ref(); }
+    void deref() const final { RefCounted::deref(); }
```
?

philn replied:
Needed, they are pure virtual in AbstractRefCounted:

```
Source/WebKit/NetworkProcess/webrtc/rice/RiceBackend.h:89:101: error: allocating an object of abstract class type 'RiceBackend'
   89 |     template<typename... Args> static Ref<RiceBackend> create(Args&&...args) { return adoptRef(*new RiceBackend(std::forward<Args>(args)...)); }
      |                                                                                                     ^
/var/home/phil/WebKit/WebKitBuild/GTK/Release/WTF/Headers/wtf/AbstractRefCounted.h:34:18: note: unimplemented pure virtual method 'ref' in 'RiceBackend'
   34 |     virtual void ref() const = 0;
      |                  ^
/var/home/phil/WebKit/WebKitBuild/GTK/Release/WTF/Headers/wtf/AbstractRefCounted.h:35:18: note: unimplemented pure virtual method 'deref' in 'RiceBackend'
   35 |     virtual void deref() const = 0;
      |                  ^
1 error generated.
```

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,783 @@
+        auto& host = result.returnValue();
+        auto address = makeString(host, ':', port);
+        GST_DEBUG_OBJECT(agent.get(), "STUN server address resolved to %s", address.ascii().data());
+        GUniquePtr<RiceAddress> stunAddress(rice_address_new_from_string(address.ascii().data()));
```
Create the CString in the stack by calling ascii(), otherwise we are duplicating the string twice. Then you just call data on the CString.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,783 @@
+
+        auto turnAddressString = makeString(result.returnValue(), ':', port);
+        GST_DEBUG_OBJECT(agent.get(), "TURN address resolved to %s", turnAddressString.ascii().data());
+        GUniquePtr<RiceAddress> turnAddress(rice_address_new_from_string(turnAddressString.ascii().data()));
```
Create the CString in the stack to avoid dupping the string twice.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,783 @@
+    if (!backend->priv->iceBackend)
+        return;
+
+    auto turnUrl = String::fromUTF8(uri);
```
When talking to Darin, I understood he had the intention of removing fromUTF8 in favor of spans. Even when this is correct, I'd suggest that for newly written code we use the `String(std::span<const char8_t>)` constructor by doing something like:

```
    auto turnUrl = String(byteCast<char8_t>(unsafeSpan(uri)));
```

I am not telling you to change this. Juist keep it in mind for the future.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,783 @@
+
+        auto newCandidateSdp = makeString(prefix, ' ', result.returnValue(), ' ', suffix);
+        GST_DEBUG("SDP for resolved address: %s", newCandidateSdp.utf8().data());
+        GUniquePtr<RiceCandidate> newCandidate(rice_candidate_new_from_sdp_string(newCandidateSdp.utf8().data()));
```
CString in the stack

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,783 @@
+    auto threadName = makeString("webrtc-rice-"_s, id);
+    counter.exchangeAdd(1);
+
+    priv->runLoop = RunLoop::create(ASCIILiteral::fromLiteralUnsafe(threadName.ascii().data()));
```
You can't do this because of lifetimes. ASCIILiterals are supposed to live forever as the underlying literal is also supposed to live forever. In this case you are creating something that is supposed to live forever from something that is going to be destroyed when the line finishes execution.

I don't know what the proper solution for this is but definitely creating a literal from something it is not a literal literal is not right.

philn replied:
I store the CStrings in a static hashset :/ Couldn't find a better solution.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,783 @@
+        auto self = weakThis.get();
+        if (!self)
+            return;
+        findStreamAndApply(self->priv->streams, streamId, [protocol, from = WTFMove(from), to = WTFMove(to), data = WTFMove(data)](const auto stream) mutable {
```
In all these calls, if not properly naming the type of the lambda parameter, at least do a `const auto*` . Here and in all calls to `findStringAndApply`.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceTransport.cpp:
```diff
@@ -0,0 +1,301 @@
+{
+    if (candidate->address) {
+        auto address = riceAddressToString(candidate->address, false);
+        gstStats->ipaddr = g_strdup(address.ascii().data());
```
I have been investigating this line to see if we could just write `CString::leakData()`  but the way CString is implemented makes this impossible without a full rework of the class. Sadly, we can't optimize this more.

See https://github.com/WebKit/WebKit/pull/54390#discussion_r2564255413 .

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,791 @@
+    if (!localAddress.address.endsWith(".local"_s)) {
+        auto errorMessage = makeString("Candidate address \""_s, localAddress.address, "\" does not end with '.local'"_s);
+        GST_ERROR_OBJECT(ice, "%s", errorMessage.utf8().data());
+        GUniquePtr<GError> error(g_error_new(GST_WEBRTC_ERROR, GST_WEBRTC_ERROR_INTERNAL_FAILURE, "%s", errorMessage.utf8().data()));
```
place the CString in the stack to not dup twice.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,791 @@
+        if (result.hasException()) {
+            auto& errorMessage = result.exception().message();
+            GST_ERROR("%s", errorMessage.utf8().data());
+            GUniquePtr<GError> error(g_error_new(GST_WEBRTC_ERROR, GST_WEBRTC_ERROR_INTERNAL_FAILURE, "%s", errorMessage.utf8().data()));
```
ditto

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,791 @@
+    static HashSet<CString> threadNames;
+    threadNames.add(threadNameCString);
+
+    priv->runLoop = RunLoop::create(ASCIILiteral::fromLiteralUnsafe(threadNameCString.data()));
```
Add a FIXME comment here about us having to abuse the ASCIILiteral::fromLiteralUnsafe because the API uses ASCIILiteral instead of String, but there should not be real problems for the time being as we're keeping the CStrings reffed so the "ASCIILiterals" are alive.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,791 @@
+        // FIXME: Convert to CStringView gymnastics.
+        auto sdp = String::fromUTF8(sdpCandidate.get());
+        ASSERT(sdp.startsWith("a="_s));
+        auto strippedSdp = sdp.substring(2);
```
```suggestion
        auto sdp = CStringView::unsafeFromUTF8(sdpCandidate.get());
        ASSERT(startsWith(sdp.span(), "a="_s));
        String strippedSdp(sdp.span().subspan(2));
```

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,791 @@
+    CStringView transport;
+    for (const auto& [key, value] : queryParameters(url)) {
+        if (key == "transport"_s) {
+            transport = CStringView::unsafeFromUTF8(value.ascii().data());
```
Now that I see this, I realize that I was wrong. Either StringView (or String if value is a String) is ok. Please for give me, it's a big patch :sweat:

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -167,10 +174,26 @@ bool GStreamerMediaEndpoint::initializePipeline()
 
-    auto binName = makeString("webkit-webrtcbin-"_s, nPipeline++);
-    m_webrtcBin = makeGStreamerElement("webrtcbin"_s, binName);
+    auto binName = makeString("webkit-webrtcbin-"_s, nPipeline);
+    nPipeline++;
+
+#if USE(LIBRICE) && GST_CHECK_VERSION(1, 20, 0)
+    auto disableNetworkSandbox = CStringView::unsafeFromUTF8(g_getenv("WEBKIT_GST_DISABLE_WEBRTC_NETWORK_SANDBOX"));
+    if (gst_check_version(1, 22, 0) && disableNetworkSandbox != "1"_s) {
+        auto peerConnectionBackend = this->peerConnectionBackend();
+        if (!peerConnectionBackend)
+            return false;
+
+        auto agent = webkitGstWebRTCCreateIceAgent(makeString(binName, ":ice"_s), peerConnectionBackend->context());
+        if (!agent) {
+            gst_printerrln("Unable to create WebRTC ICE agent");
+            return false;
+        }
+        m_webrtcBin = gst_element_factory_create_full(webrtcBinFactory.get(), "name", binName.ascii().data(), "ice-agent", agent, nullptr);
+    }
+#endif // USE(LIBRICE) && GST_CHECK_VERSION(1, 20, 0)
     if (!m_webrtcBin)
-        return false;
+        m_webrtcBin = gst_element_factory_create(webrtcBinFactory.get(), binName.ascii().data());
```
put the CString on the stack

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,791 @@
+    CStringView transport;
+    for (const auto& [key, value] : queryParameters(url)) {
+        if (key == "transport"_s) {
+            transport = CStringView::unsafeFromUTF8(value.ascii().data());
```
Again, go to StringView (or String) and please forgive me :sweat:

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -0,0 +1,791 @@
+    if (!localAddressResult.has_value()) {
+        auto errorMessage = makeString("Failed to retrieve address from candidate: "_s, localAddressResult.error().message);
+        GST_ERROR_OBJECT(ice, "%s", errorMessage.utf8().data());
+        GUniquePtr<GError> error(g_error_new(GST_WEBRTC_ERROR, GST_WEBRTC_ERROR_INTERNAL_FAILURE, "%s", errorMessage.utf8().data()));
```
place the CString in the stack

aperezdc commented on Source/cmake/FindRice.cmake:
```diff
@@ -0,0 +1,139 @@
+    find_path(RiceProto_INCLUDE_DIR
+        NAMES rice-proto.h
+        HINTS ${PC_RICE_PROTO_INCLUDEDIR}
+        ${PC_RICE_PROTO_INCLUDE_DIRS}
```
Very minor thing... typically we align the variable expansions like this:


```suggestion
        HINTS ${PC_RICE_PROTO_INCLUDEDIR}
              ${PC_RICE_PROTO_INCLUDE_DIRS}
```

(The same goes for the other instances below.)

aperezdc commented on Source/cmake/FindRice.cmake:
```diff
@@ -0,0 +1,139 @@
+        else ()
+           list(APPEND Rice_LIBS_NOT_FOUND "Proto (optional)")
+        endif ()
+    endif ()
```
Use `Rice_Proto_FOUND`, because following that naming scheme allows to use the `find_package_handle_standard_args(Rice HANDLE_COMPONENTS ...)` helper macro to have CMake do additional checks and report for each component. Also, the rest of the block is not really needed when using the helper macro.


```suggestion
        set(Rice_Proto_FOUND TRUE)
    endif ()
```

aperezdc commented on Source/cmake/FindRice.cmake:
```diff
@@ -0,0 +1,139 @@
+           list(APPEND Rice_LIBS_NOT_FOUND "Io (required)")
+        else ()
+           list(APPEND Rice_LIBS_NOT_FOUND "Io (optional)")
+        endif ()
```
Same here, we only need to set `Rice_Io_FOUND`:


```suggestion
        set(Rice_Io_FOUND TRUE)
```

aperezdc commented on Source/cmake/FindRice.cmake:
```diff
@@ -0,0 +1,139 @@
+            message(STATUS " ${found}")
+        endforeach ()
+    endif ()
+endif ()
```
Instead of this, we can use the helper macro I mentioned above:


```suggestion
include(FindPackageHandleStandardArgs)
find_package_handle_standard_args(Rice
    VERSION_VAR Rice_VERSION
    HANDLE_COMPONENTS
)
```

aperezdc commented on Source/cmake/FindRice.cmake:
```diff
@@ -0,0 +1,139 @@
+endif ()
+
+if ("Io" IN_LIST Rice_FIND_COMPONENTS)
+    pkg_check_modules(PC_RICE_IO rice-io)
```
We could set the `Rice_VERSION` variable from the information fetched with pkg-config, and then we can let `find_package_handle_standard_args()` do the version check for us:

```suggestion
    pkg_check_modules(PC_RICE_IO rice-io)

    if (NOT Rice_VERSION)
        set(Rice_VERSION ${PC_RICE_IO_VERSION})
    endif ()
```

aperezdc commented on Source/cmake/FindRice.cmake:
```diff
@@ -0,0 +1,139 @@
+find_package(PkgConfig QUIET)
+
+if ("Proto" IN_LIST Rice_FIND_COMPONENTS)
+    pkg_check_modules(PC_RICE_PROTO rice-proto)
```
We could set the `Rice_VERSION` variable from the information fetched with pkg-config, and then we can let `find_package_handle_standard_args()` do the version check for us:


```suggestion
    pkg_check_modules(PC_RICE_PROTO rice-proto)
    
    if (NOT Rice_VERSION)
        set(Rice_VERSION ${PC_RICE_PROTO_VERSION})
    endif ()
```

aperezdc reviewed (changes requested):
@philn The CMake find-module is much nicer now, but I think we should simplify it a bit using the built-in `find_package_handle_standard_arguments()` helper macro, see the comments for info on how to do that. Thanks!

aperezdc reviewed (approved):
The CMake bits look great now, thanks!

philn committed:
[GStreamer][WebRTC] Move network access to NetworkProcess
https://bugs.webkit.org/show_bug.cgi?id=244002

Reviewed by Xabier Rodriguez-Calvar and Adrian Perez de Castro.

This patch introduces a build-time dependency on librice for the GTK port. The USE_LIBRICE CMake
option can be set to FALSE if this dependency is not desired. It is FALSE by default on WPE.

When USE_LIBRICE is enabled, the WebRTC network handling is now deferred to the NetworkProcess,
bi-directional communication is performed with the WebProcess which makes use of a custom
GstWebRTCICE implementation. Runtime to the previous un-sandboxed libnice-based implementation can
be done by setting the WEBKIT_GST_DISABLE_WEBRTC_NETWORK_SANDBOX=1 environment variable.

Canonical link: https://commits.webkit.org/303623@main
